client/utils/src/notification.rs

```diff
@@ -0,0 +1,157 @@
+#[derive(Clone)]
+pub struct NotificationSender<Payload: Clone> {
+	subscribers: SharedSenders<Payload>,
+}
+
+impl<Payload: Clone> NotificationSender<Payload> {
+	/// The `subscribers` should be shared with a corresponding `NotificationStream`.
+	fn new(subscribers: SharedSenders<Payload>) -> Self {
+		Self { subscribers }
+	}
+
+	/// Send out a notification to all subscribers that a new payload is available for a
+	/// block.
+	pub fn notify<E>(&self, payload: impl FnOnce() -> Result<Payload, E>) -> Result<(), E> {
+		let mut subscribers = self.subscribers.lock();
+
+		// do an initial prune on closed subscriptions
+		subscribers.retain(|n| !n.is_closed());
+
+		if !subscribers.is_empty() {
+			let payload = payload()?;
+			subscribers.retain(|n| n.unbounded_send(payload.clone()).is_ok());
+		}
+
+		Ok(())
+	}
+}
+
+/// The receiving half of the notifications channel.
+///
+/// The `NotificationStream` entity stores the `SharedSenders` so it can be
+/// used to add more subscriptions.
+#[derive(Clone)]
+pub struct NotificationStream<Payload: Clone, TK: TracingKeyStr> {
+	subscribers: SharedSenders<Payload>,
+	_trace_key: PhantomData<TK>,
+}
+
+impl<Payload: Clone, TK: TracingKeyStr> NotificationStream<Payload, TK> {
+	/// Creates a new pair of receiver and sender of `Payload` notifications.
+	pub fn channel() -> (NotificationSender<Payload>, Self) {
+		let subscribers = Arc::new(Mutex::new(vec![]));
+		let receiver = NotificationStream::new(subscribers.clone());
+		let sender = NotificationSender::new(subscribers);
+		(sender, receiver)
+	}
+
+	/// Create a new receiver of `Payload` notifications.
+	///
+	/// The `subscribers` should be shared with a corresponding `NotificationSender`.
+	fn new(subscribers: SharedSenders<Payload>) -> Self {
+		Self { subscribers, _trace_key: PhantomData }
+	}
+
+	/// Subscribe to a channel through which the generic payload can be received.
+	pub fn subscribe(&self) -> TracingUnboundedReceiver<Payload> {
+		let (sender, receiver) = tracing_unbounded(TK::TRACING_KEY);
+		self.subscribers.lock().push(sender);
+		receiver
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use futures::{future, task::Poll, StreamExt};
+
+	#[derive(Clone)]
+	pub struct DummyTracingKey;
+	impl TracingKeyStr for DummyTracingKey {
+		const TRACING_KEY: &'static str = "test_notification_stream";
+	}
+
+	type StringStream = NotificationStream<String, DummyTracingKey>;
+
+	#[test]
+	fn notification_channel_simple() {
+		let (sender, stream) = StringStream::channel();
+
+		let test_payload = String::from("test payload");
+		let closure_payload = test_payload.clone();
+		let mut times = 0;
+		let allow_once = future::poll_fn(|_cx| {
+			times += 1;
+			if times <= 1 {
+				Poll::Pending
+			} else {
+				Poll::Ready(())
+			}
+		});
+
+		// Create a future to receive a single notification
+		// from the stream and verify its payload.
+		let future = stream.subscribe().take_until(allow_once).for_each(move |payload| {
```

Comment:
```suggestion
		let future = stream.subscribe().take(1).for_each(move |payload| {
```

Isn't this the same as we're trying to accomplish?